src/System.Text.Utf8/tests/RandomTests.cs

```diff
@@ -518,5 +518,152 @@ public unsafe void StringEquals(string text)
+            for (int i = 0; i < len; i++)
+            {
+                // 0x20 is a spacebar, writing explicitly so
+                // the value is more predictable
+                bytes[i] = unchecked((byte)(0x20 + i));
+                bytes[i + len] = bytes[i];
+            }
+
+            Utf8String sFromBytes = new Utf8String(bytes);
+            Utf8String s1FromBytes = sFromBytes.Substring(0, len);
+            Utf8String s2FromBytes = sFromBytes.Substring(len, len);
+
+            unsafe
+            {
+                fixed (byte* pinnedBytes = bytes)
+                {
+                    Utf8String sFromSpan = new Utf8String(new ByteSpan(pinnedBytes, len * 2));
+                    Utf8String s1FromSpan = sFromSpan.Substring(0, len);
+                    Utf8String s2FromSpan = sFromSpan.Substring(len, len);
+                    TestHashesSameForEquivalentString(s1FromBytes, s2FromBytes);
+                    TestHashesSameForEquivalentString(s1FromSpan, s2FromSpan);
+                    TestHashesSameForEquivalentString(s1FromSpan, s2FromBytes);
+                }
+            }
+        }
+
+        private void TestHashesSameForEquivalentString(Utf8String a, Utf8String b)
+        {
+            // for sanity
+            Assert.Equal(a.Length, b.Length);
+            Assert.Equal(a, b);
+
+            for (int i = 0; i < a.Length; i++)
+            {
+                Utf8String prefixOfA = a.Substring(i, a.Length - i);
+                Utf8String prefixOfB = b.Substring(i, b.Length - i);
+                // sanity
+                Assert.Equal(prefixOfA, prefixOfB);
+                Assert.Equal(prefixOfA.GetHashCode(), prefixOfB.GetHashCode());
+
+                // for all suffixes
+                Utf8String suffixOfA = a.Substring(a.Length - i, i);
+                Utf8String suffixOfB = b.Substring(b.Length - i, i);
+                Assert.Equal(suffixOfA, suffixOfB);
+            }
+        }
+
+        [Theory]
+        [InlineData("", "")]
+        [InlineData("abc", "")]
+        [InlineData("", "a")]
+        [InlineData("", "abc")]
+        [InlineData("", "abc")]
+        [InlineData("", "abc")]
+        [InlineData("abc", "a")]
+        [InlineData("abc", "ab")]
+        [InlineData("abc", "abc")]
+        [InlineData("abc", "b")]
+        [InlineData("abc", "bc")]
+        [InlineData("abc", "c")]
+        [InlineData("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab", "aaaaaaaab")]
+        [InlineData("abbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb", "abbbbbbbb")]
+        [InlineData("aaabaaa", "aabaaaa")]
+        [InlineData("aaabaaa", "aadaaaa")]
+        [InlineData("abababab", "bababa")]
+        [InlineData("abababab", "bbababa")]
+        [InlineData("abababab", "bababaa")]
+        [InlineData("aaabaacaaac", "aaac")]
+        [InlineData("aaabaacaaac", "aaac")]
+        [InlineData("baabaaabaaaa", "baaaa")]
+        public void IndexOfSubstringFromAndToUtf8String(string s, string substring)
```

Comment:
It seems you are testing three different things here. IndexOf(string), TrySubstringFrom and TrySubstringTo. Please, break this up into separate tests. Also, let's use some more meaningful names in the variables. s, u8s and so on are kind of bad names.